File: srni-mobile/src/services/skipLogic.ts
/**
 * Motor de skip logic offline — evaluador puro sin I/O.
 *
 * Replica exactamente la lógica del backend (EvaluarSkipLogicView):
 *   - HABILITAR: pregunta oculta por defecto; visible solo si la condición se cumple.
 *   - DESHABILITAR: pregunta visible por defecto; oculta si la condición se cumple.
 *   - OBLIGAR: hace la pregunta obligatoria (y visible).
 *   - FINALIZAR: cierra el capítulo cuando se cumple.
 *
 * Dos tipos de condición:
 *   - pregunta_origen_codigo + valor_trigger → depende de la respuesta a otra pregunta.
 *   - expresion_origen → depende del CONTEXTO de la víctima (edad, sexo, etnia, RUV).
 *     Ej: "edad >= 18 and edad <= 50", "sexo == '1'", "etnia == 'indigena'".
 *     Antes NO se evaluaba offline; ahora sí, con el ContextoVictima.
 *
 * Identificadores: código_externo (strings del diccionario V8).
 */

import type { ReglaSkipLogicRow, PreguntaRow } from '../db/instrumentoDao';

// ─── Tipos públicos ───────────────────────────────────────────────────────────

/** Mapa de respuestas: codigo_externo → valor actual ('' si sin respuesta) */
export type RespuestasMap = Record<string, string>;

/**
 * Contexto de la víctima para evaluar expresiones demográficas/étnicas offline.
 * Se arma desde VictimaResumenFuente (datos que el RUV ya tiene).
 */
export interface ContextoVictima {
  edad?: number;          // años cumplidos
  sexo?: string;          // '1'=hombre, '2'=mujer (valores del diccionario A8)
  etnia?: string;         // 'indigena' | 'negro_afro' | 'rom' | 'ninguno'
  ruvIncluido?: boolean;  // incluido en el RUV
}

export interface ResultadoSkipLogic {
  visibles: Set<string>;      // Set de codigo_externo visibles
  obligatorias: Set<string>;  // Set de codigo_externo ahora obligatorias
  finalizar: boolean;
}

// ─── Motor ────────────────────────────────────────────────────────────────────

/**
 * Determina qué preguntas son visibles dado el estado actual de respuestas
 * y el contexto de la víctima. Espejo de EvaluarSkipLogicView.post() del backend.
 */
export function calcularVisibles(
  preguntas: Pick<PreguntaRow, 'id' | 'codigo_externo' | 'obligatoria'>[],
  reglas: ReglaSkipLogicRow[],
  respuestas: RespuestasMap,
  contexto: ContextoVictima = {},
): ResultadoSkipLogic {
  const visibles = new Set<string>();
  const obligatorias = new Set<string>();
  let finalizar = false;

  for (const pregunta of preguntas) {
    const reglasEntrantes = reglas.filter(
      (r) => r.pregunta_afectada_codigo === pregunta.codigo_externo,
    );

    if (reglasEntrantes.length === 0) {
      visibles.add(pregunta.codigo_externo);
      if (pregunta.obligatoria) obligatorias.add(pregunta.codigo_externo);
      continue;
    }

    const tieneHabilitar = reglasEntrantes.some((r) => r.accion === 'HABILITAR');
    let visible = !tieneHabilitar;

    for (const regla of reglasEntrantes) {
      if (_reglaActiva(regla, respuestas, contexto)) {
        if (regla.accion === 'HABILITAR') {
          visible = true;
        } else if (regla.accion === 'DESHABILITAR') {
          visible = false;
        } else if (regla.accion === 'OBLIGAR') {
          visible = true;
          obligatorias.add(pregunta.codigo_externo);
        } else if (regla.accion === 'FINALIZAR') {
          finalizar = true;
        }
      }
    }

    if (visible) {
      visibles.add(pregunta.codigo_externo);
      if (pregunta.obligatoria) obligatorias.add(pregunta.codigo_externo);
    }
  }

  // Reglas FINALIZAR a nivel capítulo (sin pregunta_afectada): se evalúan aparte.
  for (const regla of reglas) {
    if (regla.accion === 'FINALIZAR' && !regla.pregunta_afectada_codigo) {
      if (_reglaActiva(regla, respuestas, contexto)) finalizar = true;
    }
  }

  return { visibles, obligatorias, finalizar };
}

/**
 * Descompone la respuesta en los valores seleccionados:
 *   - array JSON de multi-select ('["1","3"]') → sus elementos como strings;
 *   - cualquier otra cosa (selección única, texto, booleano) → un único valor escalar.
 * Vacío → []. Espejo de _valores_seleccionados() del backend.
 */
function _valoresSeleccionados(raw: string): string[] {
  const s = (raw ?? '').trim();
  if (!s) return [];
  if (s.startsWith('[')) {
    try {
      const arr = JSON.parse(s);
      return Array.isArray(arr) ? arr.map((x) => String(x)) : [s];
    } catch {
      return [s];
    }
  }
  return [s];
}

/** Evalúa si una regla debe dispararse. Espejo de _regla_activa() del backend. */
function _reglaActiva(
  regla: ReglaSkipLogicRow,
  respuestas: RespuestasMap,
  contexto: ContextoVictima,
): boolean {
  if (regla.pregunta_origen_codigo) {
    const valorActual = respuestas[regla.pregunta_origen_codigo] ?? '';
    if (!regla.valor_trigger) return !!valorActual;
    // Soporta origen de selección ÚNICA (escalar) y MÚLTIPLE (LISTA_MULTIPLE, que
    // guarda un array JSON: '["1","3"]'). La regla dispara si CUALQUIER valor del
    // trigger está entre los seleccionados. Compatible hacia atrás: para un escalar,
    // seleccionados = [valorActual] → equivale a la comparación anterior.
    const seleccionados = _valoresSeleccionados(valorActual);
    const triggers = regla.valor_trigger.split(',').map((v) => v.trim());
    return triggers.some((t) => seleccionados.includes(t));
  }
  // expresion_origen — se evalúa con el contexto de la víctima (edad/sexo/etnia/RUV)
  // Y, además, puede referenciar la respuesta de OTRA pregunta por su codigo_externo,
  // lo que permite condiciones AND mixtas (ej. "etnia == 'indigena' and D6 == '2'").
  if (regla.expresion_origen) {
    return _evaluarExpresion(regla.expresion_origen, contexto, respuestas);
  }
  return false;
}

// ─── Evaluador de expresiones (seguro, sin eval) ────────────────────────────────

/**
 * Evalúa expresiones simples del diccionario contra el contexto de la víctima
 * y, opcionalmente, contra las respuestas ya capturadas.
 * Soporta: and / or · operadores < <= > >= == != · variables edad, sexo, etnia,
 * ruv_incluido · valores numéricos, 'string' entre comillas, true/false ·
 * REFERENCIA a otra pregunta por su codigo_externo (cualquier nombre que no sea
 * una variable de contexto se busca en `respuestas`).
 * Ejemplos: "edad >= 18 and edad <= 50" · "sexo == '2'" · "etnia == 'indigena'"
 *           · "etnia == 'indigena' and D6 == '2'"  (AND étnico + respuesta).
 * Para orígenes multi-select (respuesta = array JSON), == / != son pertenencia.
 */
export function _evaluarExpresion(
  expr: string,
  ctx: ContextoVictima,
  respuestas: RespuestasMap = {},
): boolean {
  const e = expr.trim();
  if (!e) return false;
  // Precedencia: OR (más bajo) → AND → comparación.
  const ors = e.split(/\s+or\s+/i);
  if (ors.length > 1) return ors.some((p) => _evaluarExpresion(p, ctx, respuestas));
  const ands = e.split(/\s+and\s+/i);
  if (ands.length > 1) return ands.every((p) => _evaluarExpresion(p, ctx, respuestas));

  const m = e.match(/^(\w+)\s*(<=|>=|==|!=|<|>)\s*(.+)$/);
  if (!m) return false;
  const [, nombre, op, rawVal] = m;
  const izq = _varContexto(nombre, ctx, respuestas);
  if (izq === undefined || izq === null) return false;
  const der = _parseValor(rawVal.trim());
  // Origen multi-select: la respuesta es un array JSON ('["1","3"]') → == / != es pertenencia.
  if (typeof izq === 'string' && izq.trim().startsWith('[')) {
    const sel = _valoresSeleccionados(izq);
    const dv = String(der);
    if (op === '==') return sel.includes(dv);
    if (op === '!=') return !sel.includes(dv);
  }
  return _comparar(izq, op, der);
}

function _varContexto(
  nombre: string,
  ctx: ContextoVictima,
  respuestas: RespuestasMap = {},
): number | string | boolean | undefined {
  switch (nombre) {
    case 'edad': return ctx.edad;
    case 'sexo': return ctx.sexo;
    case 'etnia': return ctx.etnia;
    case 'ruv_incluido': return ctx.ruvIncluido;
    // Cualquier otro nombre se interpreta como codigo_externo de otra pregunta:
    // se busca su respuesta. undefined si aún no fue respondida (condición → false).
    default: {
      const v = respuestas[nombre];
      return v === undefined || v === '' ? undefined : v;
    }
  }
}

function _parseValor(raw: string): number | string | boolean {
  if (/^'.*'$/.test(raw) || /^".*"$/.test(raw)) return raw.slice(1, -1);
  if (raw === 'true') return true;
  if (raw === 'false') return false;
  const n = Number(raw);
  return Number.isNaN(n) ? raw : n;
}

function _comparar(izq: number | string | boolean, op: string, der: number | string | boolean): boolean {
  if (op === '==') return String(izq) === String(der);
  if (op === '!=') return String(izq) !== String(der);
  // comparaciones de orden: numéricas
  const a = Number(izq), b = Number(der);
  if (Number.isNaN(a) || Number.isNaN(b)) return false;
  switch (op) {
    case '<':  return a < b;
    case '<=': return a <= b;
    case '>':  return a > b;
    case '>=': return a >= b;
    default:   return false;
  }
}

// ─── Motivo legible de "no aplica" (Sprint — captura agrupada) ──────────────────
//
// En la captura agrupada, cada pregunta PERSONA se muestra UNA vez con una fila
// por miembro. Cuando la pregunta NO aplica a un miembro (no quedó visible al
// evaluar con el contexto/respuestas de ESE miembro), se muestra la fila en gris
// con un MOTIVO. Esta función deriva ese motivo a partir de las reglas entrantes,
// SIN re-evaluar visibilidad (eso lo hace calcularVisibles): aquí solo se explica.
//
// No altera el motor: es puramente descriptivo. Cubre los casos comunes
// (respuesta previa, edad, sexo, étnico). Si no logra derivar uno específico,
// devuelve un motivo genérico ("no cumple la condición").

/** Etiqueta humana para el valor de sexo del diccionario ('1'/'2'/'Hombre'…). */
function _etiquetaSexo(raw: string): string {
  const v = raw.replace(/^['"]|['"]$/g, '');
  if (v === '1' || /hombre/i.test(v)) return 'hombre';
  if (v === '2' || /mujer/i.test(v)) return 'mujer';
  if (v === '3') return 'intersexual';
  return v;
}

/** Etiqueta humana para el valor de etnia ('indigena'/'negro_afro'/'rom'/'ninguno'). */
function _etiquetaEtnia(raw: string): string {
  const v = raw.replace(/^['"]|['"]$/g, '');
  switch (v) {
    case 'indigena':   return 'indígena';
    case 'negro_afro': return 'negro/afro';
    case 'rom':        return 'Rom/gitano';
    case 'ninguno':    return 'sin pertenencia étnica';
    default:           return v;
  }
}

/** Traduce un átomo de comparación (edad/sexo/etnia/ruv) a texto legible. */
function _atomoLegible(atomo: string): string {
  const e = atomo.trim();
  // ruv_incluido (solo, sin comparación) → "incluido en el RUV"
  if (/^ruv_incluido$/i.test(e)) return 'requiere estar incluido en el RUV';
  // comparación encadenada "18 <= edad <= 49"
  const enc = e.match(/^(\d+)\s*(<=|<)\s*edad\s*(<=|<)\s*(\d+)$/i);
  if (enc) return `edad entre ${enc[1]} y ${enc[4]} años`;
  const m = e.match(/^(\w+)\s*(<=|>=|==|!=|<|>)\s*(.+)$/);
  if (!m) return e;
  const [, nombre, op, rawValRaw] = m;
  const rawVal = rawValRaw.trim();
  if (nombre === 'edad') {
    const txt: Record<string, string> = { '>=': 'edad ≥', '>': 'edad >', '<=': 'edad ≤', '<': 'edad <', '==': 'edad =', '!=': 'edad ≠' };
    return `${txt[op] ?? 'edad'} ${rawVal} años`;
  }
  if (nombre === 'sexo') {
    return op === '!=' ? `no es ${_etiquetaSexo(rawVal)}` : `es ${_etiquetaSexo(rawVal)}`;
  }
  if (nombre === 'etnia') {
    return op === '!=' ? `no es ${_etiquetaEtnia(rawVal)}` : `es ${_etiquetaEtnia(rawVal)}`;
  }
  if (nombre === 'ruv_incluido') {
    const val = rawVal.toLowerCase();
    return val === 'true' ? 'incluido en el RUV' : 'no incluido en el RUV';
  }
  return e;
}

/** Convierte una expresion_origen completa (con and/or) a texto legible. */
function _expresionLegible(expr: string): string {
  const e = expr.trim();
  if (!e) return 'no cumple la condición';
  const ors = e.split(/\s+or\s+/i);
  if (ors.length > 1) return ors.map((p) => _expresionLegible(p)).join(' o ');
  const ands = e.split(/\s+and\s+/i);
  if (ands.length > 1) return ands.map((p) => _atomoLegible(p)).join(' y ');
  return _atomoLegible(e);
}

/** Describe la condición de UNA regla en términos legibles (origen del motivo). */
function _condicionLegible(regla: ReglaSkipLogicRow): string {
  if (regla.pregunta_origen_codigo) {
    const trig = (regla.valor_trigger ?? '').trim();
    if (!trig) return `depende de ${regla.pregunta_origen_codigo}`;
    const valores = trig.includes(',')
      ? trig.split(',').map((v) => v.trim()).join(' / ')
      : trig;
    return `${regla.pregunta_origen_codigo} = ${valores}`;
  }
  if (regla.expresion_origen) return _expresionLegible(regla.expresion_origen);
  return 'no cumple la condición';
}

/**
 * Deriva el motivo legible por el que una pregunta PERSONA NO aplica a un
 * miembro. Recibe el mismo contexto/respuestas con que se evaluó visibilidad
 * para ESE miembro. Devuelve undefined si la pregunta en realidad SÍ aplica
 * (no debería llamarse en ese caso) o si no hay reglas que la oculten.
 *
 *   - Si está oculta por defecto (tiene HABILITAR pero ninguna se cumplió):
 *     motivo = la condición de la regla HABILITAR (requisito para que aparezca).
 *   - Si una DESHABILITAR se cumplió: motivo = esa condición (la que la ocultó).
 */
export function motivoOcultaPregunta(
  pregunta: Pick<PreguntaRow, 'codigo_externo'>,
  reglas: ReglaSkipLogicRow[],
  respuestas: RespuestasMap,
  contexto: ContextoVictima = {},
): string | undefined {
  const entrantes = reglas.filter(
    (r) => r.pregunta_afectada_codigo === pregunta.codigo_externo,
  );
  if (entrantes.length === 0) return undefined;

  // 1) ¿Una DESHABILITAR activa la ocultó? Ese es el motivo más directo.
  const deshabilitan = entrantes.filter((r) => r.accion === 'DESHABILITAR');
  for (const r of deshabilitan) {
    if (_reglaActiva(r, respuestas, contexto)) {
      return _condicionLegible(r);
    }
  }

  // 2) Oculta por defecto: tiene HABILITAR/OBLIGAR pero ninguna se cumplió.
  //    El motivo es el REQUISITO no satisfecho (la condición que la mostraría).
  const habilitan = entrantes.filter(
    (r) => r.accion === 'HABILITAR' || r.accion === 'OBLIGAR',
  );
  if (habilitan.length > 0) {
    const condiciones = habilitan.map((r) => _condicionLegible(r));
    // Si hay varias rutas para habilitarla, se listan como alternativas.
    const unicas = [...new Set(condiciones)];
    return `requiere ${unicas.join(' o ')}`;
  }

  return 'no cumple la condición';
}

File: srni-backend/apps/formulario/skiplogic.py
"""
Motor de skip-logic — UNA sola implementación para todo el backend.

Antes esto vivía dentro de `EvaluarSkipLogicView`, y el resto del backend que
necesitaba lo mismo se lo copiaba: `services/respuestas.py` importaba dos
helpers privados desde `views.py` y se reimplementaba el trigger, y
`SesionEncuesta.recalcular_porcentaje` directamente no evaluaba nada y contaba
TODAS las obligatorias del instrumento.

Eso último es lo que QA reportó como APK-005 —«sesión Completada con la barra en
0 %»— y no era un defecto de la barra: el denominador incluía obligatorias que
la skip-logic mantiene OCULTAS y que por lo tanto nadie puede responder nunca.
Una entrevista legítimamente terminada se guardaba en 55 %, o en 0 % si el
instrumento tenía muchas preguntas condicionales.

La regla que se sigue acá: **el criterio de visibilidad es uno solo**. Si la
vista tuviera el suyo y el modelo otro, el panel web y la APK podrían informar
distinto sobre la misma sesión, que es exactamente el tipo de contradicción que
le hace perder la confianza en el sistema a quien está en campo.

Este módulo es además el espejo de `src/services/skipLogic.ts` del móvil: los
dos tienen que decidir igual sobre los mismos datos.
"""
import ast
import json
import operator as _op

from .models import AccionSkipChoices

# ─── Evaluador AST seguro para expresiones de skip logic ─────────────────────
# Reemplaza eval() — solo permite comparaciones, operadores booleanos y literales.
# Las expresiones provienen de la BD (no de usuarios), pero usar eval() con datos
# externos es mala práctica incluso si la fuente es "confiable".

_CMP_OPS = {
    ast.Eq:    _op.eq,
    ast.NotEq: _op.ne,
    ast.Lt:    _op.lt,
    ast.LtE:   _op.le,
    ast.Gt:    _op.gt,
    ast.GtE:   _op.ge,
    ast.In:    lambda a, b: a in b,
    ast.NotIn: lambda a, b: a not in b,
}

_BOOL_OPS = {
    ast.And: all,
    ast.Or:  any,
}


class ValorDesconocido(Exception):
    """Una variable de la expresión no tiene valor conocido.

    No es un error de programación: es «todavía no sé la edad de esta persona».
    La expresión entera se resuelve como False, que es lo mismo que hace el
    móvil (`_varContexto` devuelve undefined y `_evaluarExpresion` corta).
    """


# `true`/`false`/`null` van en minúscula en las expresiones de los fixtures
# —vienen del diccionario, no de Python—, así que el AST los ve como nombres.
# Sin esto, `ruv_incluido == false` compararía contra un valor desconocido.
_LITERALES = {'true': True, 'false': False, 'null': None, 'none': None}


def _safe_eval(node: ast.AST, ctx: dict):
    """Evalúa un nodo AST restringido a comparaciones y literales."""
    if isinstance(node, ast.Expression):
        return _safe_eval(node.body, ctx)
    if isinstance(node, ast.Constant):
        return node.value
    if isinstance(node, ast.Name):
        if node.id in _LITERALES:
            return _LITERALES[node.id]
        # Ausente o vacío = desconocido. Antes esto devolvía '', y ese default
        # silencioso decidía mal en la dirección peligrosa: `etnia != 'ninguno'`
        # daba VERDADERO contra el vacío, así que preguntas del capítulo étnico
        # quedaban exigidas a todo el mundo y ninguna sesión llegaba nunca al
        # 100 %. Desconocido tiene que significar «no sé», no «distinto».
        if node.id not in ctx:
            raise ValorDesconocido(node.id)
        valor = ctx[node.id]
        if valor is None or (isinstance(valor, str) and not valor.strip()):
            raise ValorDesconocido(node.id)
        return valor
    if isinstance(node, ast.List):
        return [_safe_eval(el, ctx) for el in node.elts]
    if isinstance(node, ast.Compare):
        left = _safe_eval(node.left, ctx)
        for cmp_op, comparator in zip(node.ops, node.comparators):
            fn = _CMP_OPS.get(type(cmp_op))
            if fn is None:
                raise ValueError(f'Operador no permitido: {cmp_op}')
            if not fn(left, _safe_eval(comparator, ctx)):
                return False
            left = _safe_eval(comparator, ctx)
        return True
    if isinstance(node, ast.BoolOp):
        fn = _BOOL_OPS.get(type(node.op))
        if fn is None:
            raise ValueError(f'Operador booleano no permitido: {node.op}')
        values = [_safe_eval(v, ctx) for v in node.values]
        return fn(values)
    if isinstance(node, ast.UnaryOp) and isinstance(node.op, ast.Not):
        return not _safe_eval(node.operand, ctx)
    raise ValueError(f'Nodo AST no permitido: {type(node).__name__}')


def evaluar_expresion_segura(expresion: str, contexto: dict, respuestas: dict = None) -> bool:
    """Evalúa una expresión de skip logic de forma segura usando AST.

    El ámbito combina las respuestas ya capturadas (indexadas por codigo_externo)
    como base y el contexto de la víctima (edad/sexo/etnia/ruv_incluido) con
    precedencia. Así una expresión puede referenciar la respuesta de OTRA pregunta
    por su código, habilitando condiciones AND mixtas
    (ej. "etnia == 'indigena' and D6 == '2'"). Espejo de _evaluarExpresion (móvil).
    Convención: los valores del lado derecho van entre comillas ('2', 'true').

    Ante cualquier problema devuelve False —variable desconocida, tipos que no
    se comparan, expresión mal escrita—. Es deliberado: una regla que no se
    puede evaluar NO se dispara, así que en el peor caso la pregunta se comporta
    como si su condición no se cumpliera, en vez de reventar la captura en campo.
    """
    try:
        tree = ast.parse(expresion, mode='eval')
        ambito = dict(respuestas or {})
        ambito.update(contexto or {})
        return bool(_safe_eval(tree, ambito))
    except Exception:
        return False


# ─── Motor declarativo ───────────────────────────────────────────────────────

def valores_seleccionados(valor: str) -> list:
    """Descompone la respuesta en los valores seleccionados.

    Array JSON de un multi-select ('["1","3"]') → sus elementos como strings;
    cualquier otra cosa (selección única, texto, booleano) → un único valor escalar.
    Vacío → []. Espejo de _valoresSeleccionados() del móvil (skipLogic.ts).
    """
    s = (valor or "").strip()
    if not s:
        return []
    if s.startswith("["):
        try:
            arr = json.loads(s)
            if isinstance(arr, list):
                return [str(x) for x in arr]
        except (ValueError, TypeError):
            pass
        return [s]
    return [s]


def regla_activa(regla, respuestas: dict, contexto: dict = None) -> bool:
    """¿Esta regla se dispara con las respuestas y el contexto actuales?

    `respuestas` va indexado por `codigo_externo`, no por id: es la clave con la
    que están escritas las reglas y la que usa el móvil.
    """
    contexto = contexto or {}

    if regla.pregunta_origen_id:
        valor_actual = respuestas.get(regla.pregunta_origen.codigo_externo, "")
        if not regla.valor_trigger:
            return bool(valor_actual)
        # Soporta origen de selección ÚNICA (escalar) y MÚLTIPLE (LISTA_MULTIPLE,
        # que guarda un array JSON: '["1","3"]'). Dispara si CUALQUIER valor del
        # trigger está entre los seleccionados. Compatible hacia atrás: para un
        # escalar, seleccionados == [valor_actual].
        seleccionados = valores_seleccionados(valor_actual)
        triggers = [v.strip() for v in regla.valor_trigger.split(",")]
        return any(t in seleccionados for t in triggers)

    if regla.expresion_origen:
        return evaluar_expresion_segura(regla.expresion_origen, contexto, respuestas)

    return False


def calcular_visibles(preguntas, reglas, respuestas: dict, contexto: dict = None):
    """Decide qué preguntas están visibles y cuáles son obligatorias.

    Reglas de visibilidad (idénticas a `calcularVisibles` del móvil):
      - Sin reglas entrantes → visible.
      - Con alguna regla HABILITAR entrante → oculta por defecto; se muestra solo
        si esa regla se dispara.
      - DESHABILITAR activa → oculta.
      - OBLIGAR activa → visible Y obligatoria, aunque la pregunta no lo fuera.
      - FINALIZAR activa → cierra el capítulo.

    Devuelve `(visibles, obligatorias, finalizar)` con los dos primeros como
    conjuntos de `codigo_externo`.
    """
    contexto = contexto or {}
    visibles = set()
    obligatorias = set()
    finalizar = False

    # Un índice por pregunta afectada evita recorrer TODAS las reglas por cada
    # pregunta: con ~300 preguntas y ~200 reglas eso era 60.000 comparaciones en
    # un método que corre en cada respuesta guardada.
    entrantes = {}
    for r in reglas:
        if r.pregunta_afectada_id:
            entrantes.setdefault(r.pregunta_afectada_id, []).append(r)

    for pregunta in preguntas:
        reglas_entrantes = entrantes.get(pregunta.pk, [])

        if not reglas_entrantes:
            visibles.add(pregunta.codigo_externo)
            if pregunta.obligatoria:
                obligatorias.add(pregunta.codigo_externo)
            continue

        tiene_habilitar = any(
            r.accion == AccionSkipChoices.HABILITAR for r in reglas_entrantes
        )
        visible = not tiene_habilitar

        for regla in reglas_entrantes:
            if regla_activa(regla, respuestas, contexto):
                if regla.accion == AccionSkipChoices.HABILITAR:
                    visible = True
                elif regla.accion == AccionSkipChoices.DESHABILITAR:
                    visible = False
                elif regla.accion == AccionSkipChoices.OBLIGAR:
                    visible = True
                    obligatorias.add(pregunta.codigo_externo)
                elif regla.accion == AccionSkipChoices.FINALIZAR:
                    finalizar = True

        if visible:
            visibles.add(pregunta.codigo_externo)
            if pregunta.obligatoria:
                obligatorias.add(pregunta.codigo_externo)

    # Reglas FINALIZAR a nivel capítulo (sin pregunta_afectada): se evalúan aparte.
    for regla in reglas:
        if regla.accion == AccionSkipChoices.FINALIZAR and not regla.pregunta_afectada_id:
            if regla_activa(regla, respuestas, contexto):
                finalizar = True

    return visibles, obligatorias, finalizar
